### PR TITLE
N+1 Problem Resolution and Integration Test Implementation

### PR DESCRIPTION
**Pull request by lily4178993**
## Description:

In collaboration with my partner @Ayokunnumi1, we have addressed the N+1 problem across various controllers, with a particular focus on resolving the issue in the `posts#index` action. We ensured that all posts and their associated comments are fetched efficiently, eliminating unnecessary database queries.

### N+1 Problem Resolution:

- Ensured efficient loading of all posts and their associated comments for a user.
- Deliberately avoided the use of the Bullet gem for analysis, preferring to inspect the logs for performance improvements.
Below are visual comparisons:
**Before:**
image
**After:**
image

### Integration Test Implementation:

We have incorporated Capybara to create comprehensive integration tests for essential views within the project. The tests are designed to validate the functionality and flow of each page, ensuring a seamless user experience.

#### User Index Page:

- Verified that the username of all users is displayed.
- Checked for the presence of profile pictures for each user.
- Confirmed the visibility of the number of posts written by each user.
- Implemented redirection to a user's show page upon clicking on their profile.

#### User Show Page:

- Confirmed the display of the user's profile picture.
- Ensured visibility of the user's username.
- Checked the accuracy of the displayed post count for the user.
- Verified the visibility of the user's bio.
- Confirmed the display of the user's first three posts.
- Implemented a button for viewing all of a user's posts.
- Validated redirection to a post's show page upon clicking a user's post.
- Verified redirection to the user's post index page when viewing all posts.

#### User Post Index Page:

- Confirmed the visibility of the user's profile picture.
- Ensured the display of the user's username.
- Checked the accuracy of the displayed post count for the user.
- Validated the display of a post's title.
- Confirmed visibility of some of the post's body.
- Checked the display of the first comments on a post.
- Verified the accuracy of the displayed comment and like counts.
- Implemented a section for pagination for an optimal viewing experience.
- Validated redirection to a post's show page upon clicking a post.

#### Post Show Page:

- Verified the display of the post's title.
- Ensured visibility of the post's author.
- Checked the accuracy of the displayed comment and like counts.
- Confirmed the display of the post body.
- Checked the visibility of the username for each commenter.
- Validated the display of each comment made on the post.

These changes collectively enhance the project's performance and provide a more robust testing framework for future development.

## Additional Notes:

Please review the changes made and confirm that they align with the intended enhancements and fulfill the specified integration test scenarios. If further adjustments are needed, we are open to feedback for continuous improvement.

### CODE REVIEW

File: spec/views/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    expect(rendered).to have_content(user.bio)
+    expect(rendered).to have_css("img[src*='#{user.photo}']")
+    expect(rendered).to have_content("Number of posts: #{user.posts_counter}")
+  end
+
+  it 'displays user posts' do
+    posts.each do |post|
+      expect(rendered).to have_content(post.title)
+      expect(rendered).to have_content(post.text) # Assuming 'text' is the attribute for post body
+    end
```
Comment (Olamarx) [inline spec/views/user_show_spec.rb]: - [ ] While you have done a good job so far, your tests for this file are not yet complete, kindly add the pointed ones in the image
image

File: spec/views/post_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  scenario 'User sees pagination section on index page' do
+    3.times { Post.create(author: user, title: 'Test Post', text: 'This is a test post.') }
+
+    visit user_posts_path(user)
+
+    expect(page).to have_css('.post-container', count: 2)
+  end
+
+  scenario 'User redirects to post show page when clicking on a post' do
```
Comment (Olamarx) [inline spec/views/post_index_spec.rb]: - [ ] Same for this file, the test is yet to be complete as pointed in the image 
image
image

**Review by Olamarx — CHANGES_REQUESTED** [review]
Hi @lily4178993, 

Good job so far, indeed you have spent time on this project!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights
- Linters passed ✔️
- Readme file professional ✔️
- N+1 problem fixed ✔️

## Required Changes ♻️
_Check the comments under the review._
- [ ] I noticed some of your tests are failing, kindly fix it up as seen in the image 
image

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by lily4178993** [comment]
Hi @Olamarx , Thank you for the review and feedbacks.

## Changes made:
- Refactored integration specs files
- Added additonal tests

screenshot of the tests passing:
image

File: spec/views/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    expect(page).to have_content("Number of posts: #{user.posts_counter}")
+  end
+
+  scenario 'User sees post details on index page' do
+    expect(page).to have_content(Post.first.title)
+    expect(page).to have_content(Post.first.text)
+    expect(page).to have_content(Post.first.comments_counter)
+    expect(page).to have_content(Post.first.likes_counter)
+    expect(page).to have_content(Comment.first.text)
+  end
```
Comment (omar-muhamad) [inline spec/views/post_index_spec.rb]: - [ ] As the previous reviewer asked, this test should ensure you can see the first comments (more than one comment) for a user post, not only see the first comment as you did. Please fix that.

File: spec/views/post_index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+  scenario 'User sees pagination section on index page to see more posts' do
+    expect(page).to have_css('.post-container', count: 2)
+  end
```
Comment (omar-muhamad) [inline spec/views/post_index_spec.rb]: - [ ] As the previous reviewer asked, This test should ensure that the user sees the pagination section on the index page to see more posts, not see if the `div` which has a class `post-container` exists. If you want to make it the correct way you should test the existence of an element that has the `pagination-container` class according to your `post_index` view.

image

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @lily4178993,

#### Nice job making the required changes. Unfortunately, you missed some of them but you are almost there.



## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below ⬇️.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @lily4178993, and @Ayokunnumi1👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 🎉
- Good job fixing n+1 problem. ✔️
- Good job with integration specs. ✔️
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- No linter errors. :heavy_check_mark:
- PR has a good title and summary. ✔️



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by lily4178993** [comment]
Thank you @Omar-Muhamad, @thecodechaser and @Olamarx for dedicating time to review my pull request. Many thanks for all your valuable feedback.

**Pull request merged**